docs/02.API-REFERENCE.md

```diff
@@ -8065,6 +8084,164 @@ best-practice example.
 - [jerry_object_native_info_t](#jerry_object_native_info_t)
 
 
+## jerry_native_pointer_init_references
+
+**Summary**
+
+Initialize the references stored in a buffer pointed by a native pointer.
+The references are initialized to undefined. This function must be called
+before the buffer is attached to an object by
+[jerry_set_object_native_pointer](#jerry_set_object_native_pointer).
+
+*Note*:
+  - The description of [jerry_object_native_info_t](#jerry_object_native_info_t)
+    provides detailed information about these references.
+
+**Prototype**
+
+```c
+void
+jerry_native_pointer_init_references (void *native_pointer_p, uint32_t number_of_references);
+```
+
+- `native_pointer_p` - native pointer to a buffer.
+- `number_of_references` - number of references.
+
+*New in version [[NEXT_RELEASE]]*.
+
+**Example**
+
+[doctest]: # ()
+
+```c
+#include <stdlib.h>
+#include "jerryscript.h"
+
+typedef struct
+{
+  jerry_value_t a;
+  jerry_value_t b;
+  uint32_t user_value;
+} user_buffer_t;
+
+static void
+native_references_free_callback (void *native_p) /**< native pointer */
+{
+  /* References must be finalized when a buffer is no longer attached. */
+  jerry_native_pointer_finalize_references (native_p, 2);
+  free (native_p);
+} /* native_references_free_callback */
+
+static const jerry_object_native_info_t native_info =
+{
+  .free_cb = native_references_free_callback,
+  .number_of_references = 2
+};
+
+int
+main (void)
+{
+  jerry_init (JERRY_INIT_EMPTY);
+
+  jerry_value_t object_value = jerry_create_object ();
+
+  user_buffer_t *buffer_p = (user_buffer_t *) malloc (sizeof (user_buffer_t));
+
+  /* References must be initialized before a buffer is attached. */
+  jerry_native_pointer_init_references ((void *) buffer_p, 2);
+
+  jerry_set_object_native_pointer (object_value, (void *) buffer_p, &native_info);
+
+  /* References can be modified after the buffer is attached.
+   * This example sets a self reference. */
+  jerry_native_pointer_set_reference (&buffer_p->a, object_value);
+
+  jerry_release_value (object_value);
```

Comment:
What if we use the `native_info` to access the number of references for the init/finalize/set_ref methods? That way the number_of_references is only specified at a single place. The rationale is that the buffer's format is tied to the native_info struct's number_of_references member and by providing the native_info for these methods, it would be harder to miss type the number_of_refs for the init/finalize method.

For example:
```c
jerry_native_pointer_init_references (&native_info, (void *) buffer_p);
   ...
jerry_native_pointer_finalize_references (&native_info, native_p);
```

Related to this the idea is to also provide the native info, buffer start pointer, index for the ref to change and the new value for the set_ref method:
```c
jerry_native_pointer_set_reference (&native_info, (void*)buffer_p, 0, object_value);
```

Or another idea would be:
```c
jerry_native_pointer_set_reference (target_object, &native_info, 0, new_ref_value);
```
Where internally the target_object's native_info will be accessed and based on the provided index a new_ref_value will be set. However with this there would be too many native pointer access calls internally and would potentially be slower.

What do you think?

Reply:
Unfortunately you don't get the native info in the finalization callback, so passing it might be hard.

In the other case, you need to manually count the variables, and that is difficult. If you have a struct like:

typedef struct {
  jerry_value_t something;
  jerry_value_t my_val;
  jerry_value_t object_ref;
}
```

For me `&buffer_p->object_ref` is more descriptive than `(buffer_p, 2)`. You can also reorganize the values freely in the former case. The latter can only work if you use an array and constants for indexing. But that is a lot of definition sperad accross the code.